geopandas/geodataframe.py

```diff
@@ -276,6 +285,11 @@ def set_geometry(self, col, drop=False, inplace=False, crs=None):
                 raise ValueError("Unknown column %s" % col)
             except Exception:
                 raise
+            # specially check shape of level for consistency (shapely raises a
+            # TypeError but pygeos raises a ValueError)
+            if hasattr(level, "ndim") and level.ndim != 1:
```

Comment:
If you drop the `.values` on the line above where `level` is created, then `level` should be a Series instance (and not DataFrame, which you would get with a duplicated column name), making this check a bit easier IMO.